Speaker Feedback: Add REST endpoint for updating feedback with specific meta values

coreymckrill opened the pull request:
Adds the ability to mark a feedback comment as "helpful" via a new API endpoint. This is stored as comment meta, just like the other values that make up the original feedback submission. However, "helpful" can only be set during feedback update, not creation. This is achieved by introducing a "context" for each possible meta field.

This also makes some changes to the way request objects are processed by our endpoint controller to allow for the flexibility to consider the context of the request.

Refs #342 

### How to test the changes in this Pull Request:

1. To make is easier to send authenticated requests to the endpoint via something like [Insomnia](https://insomnia.rest/), you'll probably want to temporarily install the [Basic Auth plugin](https://github.com/WP-API/Basic-Auth).
1. Create some feedback comments on one or more session posts.
1. To mark one of the comments as helpful, you'll need the ID of the comment.
1. The endpoint URL will be `[your-test-site].wordcamp.org/wp-json/wordcamp-speaker-feedback/v1/feedback/[comment ID]`
1. Currently there is only valid parameter for the payload:
    * `meta[helpful]` (A boolean-ish value)
1. The response to a successful request to the endpoint should be a 201 status and an empty array.
1. Try sending a non-boolean value. You should get an error message.
1. Try including other existing meta values besides `helpful` (`rating` or `q1` for example). These should be ignored and not changed during a successful request.
1. Try including `meta[helpful]` when sending a request to [create a new feedback](https://github.com/WordPress/wordcamp.org/pull/353). It should also be ignored and not be present for the newly created comment.
1. Try authenticating as a non-admin/editor user who is also not a speaker on the session. You should get an error message.

coreymckrill committed:
Add REST endpoint for updating feedback with specific meta values

Adds the ability to mark a feedback comment as "helpful" via an API
endpoint. This is stored as comment meta, just like the other values
that make up the original feedback submission. However, "helpful"
can't be set during feedback creation. This is achieved by introducing
a "context" for each possible meta field.

coreymckrill committed:
Update tests

coreymckrill committed:
Fix tests

ryelle commented on public_html/wp-content/plugins/wordcamp-speaker-feedback/includes/comment-meta.php:
```diff
@@ -124,8 +133,29 @@ function get_feedback_meta_field_schema( $key = '' ) {
+			'sanitize_callback' => 'wp_validate_boolean',
+			'show_in_rest'      => false,
+			'default'           => false,
+			'required'          => false,
```
Should this be required? I can call the update endpoint with `meta[q1]=fake` and it returns 201, even though it doesn't update anything.

coreymckrill replied:
Ah, good point. I hesitate to make it `required` because semantically it implies that all feedback should have that value, when that is not the case. However, in 4bcda31e I made it so that if there's nothing left after meta validation callback strips out invalid keys (like `q1` in your case above, which is invalid in the update context), it will throw an error instead of returning 201. Do you think that's sufficient?

ryelle replied:
Yeah, that looks good 👍

ryelle reviewed (approved):
Left one question, but it's not a blocker.

Another issue I noticed is that the speaker needs to be a user on the site for the cap check to work, my guess is that they don't have `read` caps. Once I added the user to the site, everything worked as expected. This can also be addressed in a followup issue/PR.

coreymckrill committed:
Return error if update payload is empty

The meta validation callback strips out parameters that aren't valid
for the current context. This can cause the `comment_meta` part of the
payload to exist, but be empty. With this, `comment_meta` must actually
contain a valid key/value pair.

coreymckrill commented:
> Another issue I noticed is that the speaker needs to be a user on the site for the cap check to work, my guess is that they don't have read caps. Once I added the user to the site, everything worked as expected.

Thanks, that should definitely be addressed. I've added #425 to look into it further.
